docs/modules/ROOT/pages/how-to-guides/proc_prefetching-dependencies-to-support-hermetic-build.adoc

```diff
@@ -0,0 +1,193 @@
+
+. Navigate to your component's source code.
+. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
+. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.
+
++
+[source,yaml]
+----
+param:
+    -  name: prefetch-input
+       value: gomod <1>
+----
+<1> If the package manager lock file is located in the root directory of your project, enter `gomod`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in JSON format. For example,
+{"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}
+
+include::../partials/con_hermetic_verification.adoc[]
+
+== Enabling prefetch builds for `pip`
+Cachi2 supports pip by parsing of `pip` requirements files, including but not limited to *requirements.txt* files placed in the root of your repository. By generically parsing `pip` requirements files, it downloads the specified dependencies.
+
+IMPORTANT: The requirements file can have a different name, as you can use multiple files to provide the dependencies. These requirements files function as lockfiles, encompassing all the transitive dependencies. It's crucial to actively pin each transitive dependency listed in the requirements file to a specific version.
+
+
+s, which include all the transitive dependencies. Each transitive dependency in the *requirements.txt* file must be pinned to a specific version.
+
+.Prerequisites
+* You have an upgraded build pipeline.
+* You have a <filename> file in your repository that lists all the the dependencies.
+
+.Procedures
+For each component that you want to build hermetically:
+
+. Navigate to your component’s source code.
+. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
+. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.
+
++
+[source,yaml]
+----
+param:
+    -  name: prefetch-input
+       value: pip <1>
+----
+<1> If the package manager lock file is located in the root directory of your project, enter `pip`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in the JSON format. For example,
+{"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}
+
+. Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.
+
++
+[source,metadata]
+----
+[metadata]
+name = "my_package"
+version = "0.1.0"
+----
+
++
+NOTE: Instead of a *pyproject.toml* file, you can also create a *setup.py* or *setup.cfg* file. For information on the metadata of these files, see link:https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#project-metadata[Project Metadata].
+
+. To generate a fully resolved requirements.txt file, run
```

Comment:
Does it matter what kind of operating system they're on, or pip version they have, to be able to run this command? Do they need to add the generated file to their source repo, if so, where? the root? Do they have to generate this file? There isn't much context here, and requirements.txt wasn't referenced earlier in this procedure (other than listing that it could be processed at the start). It probably makes more sense to developers though.

Reply:
> Does it matter what kind of operating system they're on

It does. The environment where the user runs this should be as close as possible to the environment in the container build. Same OS, same python version - that should usually be enough.

> Does it matter pip version they have

pip-compile is a separate tools from pip, it comes from the [pip-tools](https://github.com/jazzband/pip-tools) package. The package does [runtime-depend on pip](https://github.com/jazzband/pip-tools/blob/6d06be8dbd88298a9e62df1415a8da28a14eadc0/pyproject.toml#L42), but `pip install pip-tools` should do the right thing.

Reply:
I had a heck of a time with the CATALOG when updating the requirements from a mac -- I had to manually revert a deletion every time because _my pip-compile_ dropped a linux requirement which was needed for CI.

Reply:
@mattreid @chmeliik  Do you think adding a Note just after the procedure in line # 90 is good?

The updated procedures will look like this:

> ...
> . Generate a fully resolved *requirements.txt* file that contains all the transitive dependencies and pins them to a specific version and hash by using the following command:
> 
> +
> [source,command]
> ----
> $ pip-compile pyproject.toml --generate-hashes
> ----

> NOTE(BULLET2):  To successfully run the command mentioned above your environment must be as close as possible to the environment in the container build. That is, the environment should have the same operating system and the same Python $major.$minor version.

> Add the *requirements.txt* to the root of your component’s source code. 

> ....

Reply:
That seems better to me. You could consider if it's worth calling out that they'd need pip-tools installed, but maybe that's obvious and pip would help them realize and install that if they didn't anyway.

Reply:
@chmeliik If they don't have pyproject.toml, but have either setup.py or setup.cfg, can they use that to generate requirements.txt? or to do the generation, they need the toml file? From the description here, all 3 sound equivalent for cachi2.

Reply:
pip-compile does support all of them and cachi2 only cares about the generated requirements.txt

* https://pip-tools.readthedocs.io/en/latest/#requirements-from-pyproject-toml
* https://pip-tools.readthedocs.io/en/latest/#requirements-from-setup-py-and-setup-cfg
* https://pip-tools.readthedocs.io/en/latest/#requirements-from-requirements-in

_N.B. cachi2 does also look at pyproject.toml, setup.py and setup.cfg, but only to figure out the name and version of your project. For the list of dependencies, cachi2 always uses requirements.txt._